Build a regression model.

In [2]:
# imports
import pandas as pd
import statsmodels.api as sm

grouped_business_df = pd.read_csv('grouped_business.csv', index_col = 0)

In [9]:
grouped_business_df

Station Name  Average Rating  Average Reviews  \
0            Broadway Residence Hall        4.175000       572.050000   
1                Casey Eye Institute        4.175000        80.850000   
2              Cleveland High School        4.300000       561.050000   
3                         Cully Park        4.083333        47.111111   
4    Doernbecher Children's Hospital        4.175000        48.750000   
..                               ...             ...              ...   
233      SW Yamhill at Director Park        4.175000      1497.900000   
234                    Shattuck Hall        4.175000       605.500000   
235         Tilikum West at SW Moody        4.025000       285.600000   
236     Urban Center Plaza at SW 6th        4.150000       741.550000   
237        W Burnside at SW Park Ave        4.200000      1707.550000   

     Number of Bikes  
0                  4  
1                  6  
2                  9  
3                 10  
4                  6  
..               ...  
233               20  
234                3  
235               18  
236               14  
237                9  

[238 rows x 4 columns]

Data audit and cleaning

In [6]:
grouped_business_df.shape
#This table has 238 rows and 4 columns

(238, 4)

In [4]:
grouped_business_df.head()

Station Name  Average Rating  Average Reviews  \
0          Broadway Residence Hall        4.175000       572.050000   
1              Casey Eye Institute        4.175000        80.850000   
2            Cleveland High School        4.300000       561.050000   
3                       Cully Park        4.083333        47.111111   
4  Doernbecher Children's Hospital        4.175000        48.750000   

   Number of Bikes  
0                4  
1                6  
2                9  
3               10  
4                6

In [7]:
#Checking for missing values
missing_values =grouped_business_df.isnull().sum()
print(missing_values)
#No missing values in the dataframe

Station Name       0
Average Rating     0
Average Reviews    0
Number of Bikes    0
dtype: int64


In [8]:
#Basic descriptive statistics for each predictor variable by Number of Bikes
descriptive_stats = grouped_business_df.groupby('Number of Bikes').describe()
descriptive_stats

Average Rating                                                 \
                         count      mean       std       min      25%     50%   
Number of Bikes                                                                 
1                          1.0  4.275000       NaN  4.275000  4.27500  4.2750   
2                          4.0  4.306250  0.082601  4.250000  4.25000  4.2750   
3                         19.0  4.196053  0.148666  3.875000  4.13750  4.2000   
4                         11.0  4.172727  0.262311  3.450000  4.13750  4.2250   
5                         26.0  4.238462  0.122929  3.950000  4.13125  4.2750   
6                         20.0  4.205000  0.200755  3.675000  4.15625  4.2500   
7                         16.0  4.204688  0.163864  3.850000  4.15000  4.2250   
8                         18.0  4.156944  0.265934  3.225000  4.15625  4.2125   
9                         24.0  4.273177  0.177921  3.925000  4.19375  4.3125   
10                         9.0  4.242593  0.091709  4.083333  4.17500  4.3000   
11                        10.0  4.150000  0.288434  3.375000  4.17500  4.2250   
12                        16.0  4.251563  0.088727  4.000000  4.21250  4.2750   
13                         6.0  4.270833  0.108877  4.125000  4.20625  4.2625   
14                         9.0  4.186111  0.155680  3.825000  4.15000  4.2250   
15                         2.0  4.187500  0.123744  4.100000  4.14375  4.1875   
16                         9.0  4.208333  0.094373  4.025000  4.20000  4.2250   
17                        11.0  4.215909  0.091017  4.000000  4.18750  4.2250   
18                        22.0  4.246591  0.115288  4.025000  4.15625  4.2750   
20                         2.0  4.162500  0.017678  4.150000  4.15625  4.1625   
21                         1.0  4.275000       NaN  4.275000  4.27500  4.2750   
23                         1.0  4.100000       NaN  4.100000  4.10000  4.1000   
25                         1.0  4.050000       NaN  4.050000  4.05000  4.0500   

                                Average Reviews                           \
                     75%    max           count         mean         std   
Number of Bikes                                                            
1                4.27500  4.275             1.0   235.950000         NaN   
2                4.33125  4.425             4.0   604.862500  322.491177   
3                4.32500  4.450            19.0   435.955263  281.167948   
4                4.30000  4.425            11.0   688.450000  275.693302   
5                4.31875  4.450            26.0   368.319231  251.454135   
6                4.33750  4.475            20.0   492.105000  467.245930   
7                4.35000  4.375            16.0   311.134375  182.454284   
8                4.26250  4.475            18.0   338.947222  222.441728   
9                4.37500  4.625            24.0   409.478125  456.402911   
10               4.30000  4.325             9.0   818.512346  468.401795   
11               4.28750  4.400            10.0   557.690000  473.908544   
12               4.30625  4.350            16.0   687.168750  352.336818   
13               4.33750  4.425             6.0   735.116667  560.679342   
14               4.30000  4.325             9.0   701.416667  446.788092   
15               4.23125  4.275             2.0   600.075000  542.174124   
16               4.22500  4.375             9.0  1096.022222  506.355752   
17               4.27500  4.350            11.0   850.700000  455.628257   
18               4.35000  4.400            22.0   792.195455  406.076819   
20               4.16875  4.175             2.0  1504.525000    9.369165   
21               4.27500  4.275             1.0   334.050000         NaN   
23               4.10000  4.100             1.0   650.900000         NaN   
25               4.05000  4.050             1.0   770.200000         NaN   

                                                                       
                 

In [10]:
X = grouped_business_df['Number of Bikes']
Y = grouped_business_df['Average Rating']

In [11]:
X = sm.add_constant(X)
lin_model = sm.OLS(Y, X).fit()

In [12]:
model = lin_model.summary()

Provide model output and an interpretation of the results. 

In [13]:
print(model)

                            OLS Regression Results                            
Dep. Variable:         Average Rating   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.004
Method:                 Least Squares   F-statistic:                   0.02730
Date:                Tue, 12 Dec 2023   Prob (F-statistic):              0.869
Time:                        22:17:59   Log-Likelihood:                 88.758
No. Observations:                 238   AIC:                            -173.5
Df Residuals:                     236   BIC:                            -166.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               4.2146      0.023    1

In [ ]:
#Inference from above statsmodel:
#R-squared measures the proportion of the variance in the dependent variable (Average Rating) explained by the independent variable(s) (Number of Bikes).
#In this case, R-squared is 0.000, indicating that the model does not explain a significant amount of variability in 'Average Rating'. The model doesn't fit the data well and is not statistically significant
#Also the p-value for the Number of Bikes is actually very high at 0.869, we typically are looking for less than 0.05.

In [15]:
#using 'Average Reviews' for the stats model
X = grouped_business_df['Number of Bikes']
Y = grouped_business_df['Average Reviews']

X = sm.add_constant(X)
lin_model = sm.OLS(Y, X).fit()
model = lin_model.summary()
print(model)

                            OLS Regression Results                            
Dep. Variable:        Average Reviews   R-squared:                       0.137
Model:                            OLS   Adj. R-squared:                  0.133
Method:                 Least Squares   F-statistic:                     37.43
Date:                Tue, 12 Dec 2023   Prob (F-statistic):           3.90e-09
Time:                        22:33:09   Log-Likelihood:                -1760.3
No. Observations:                 238   AIC:                             3525.
Df Residuals:                     236   BIC:                             3532.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const             274.1422     55.022     

Interpretation of Results:

R-squared: From the model above, an R-squared value of 0.137 indicates that about 13.7% of the variability in the dependent variable (Average Reviews) is explained by the variability in the independent variable (Number of Bikes). Thus, this model suggests a moderate level of variation in the dependent variable.

P-value: In this model output, the P-value for the Number of Bikes is 0.000. The low p-value suggests that we can reject the null hypothesis. The p-value indicates that the 'Number of Bikes' variable is a significant predictor of 'Average Reviews' in the model. 'Number of Bikes' variable has statistically significant effect on 'Average Reviews'.

Coefficient: For this linear regression model, the coefficient for Number of Bikes is approximately 30.8923, suggesting an estimated change in 'Average Reviews' for a one-unit increase in the 'Number of Bikes'.

F-statistic: In this model output, the F-statistic is 37.43 with a very low p-value (3.90e-09) which is an indicator that the model is statistically significant, and the 'Number of Bikes' is a significant predictor of 'Average Reviews'.

Conclusion: While we can establish a statistically significant relationship between the independent and dependent variables, it is also worthy to note that the relationship between them does not imply causation.


# Stretch

How can you turn the regression model into a classification model?

To convert the regression model into a classification model, I will first define which attribute I am trying to predict. Next, I'll work on grouping my continuous variables into discrete labels to help create the class. Then I'll select an approrpiate classification algorithm and afterwards, evaluate my model.